### ISSUE TITLE
resample_spec problems/questions for NIRSpec MOS

### ISSUE THREAD

**Issue by stscijgbot** [spacetelescope/jwst#2499]
Issue [JP-369](https://jira.stsci.edu/browse/JP-369) was created by James Muzerolle:

When processing NIRSpec MOS data, the products from the resample_spec step have a few peculiarities.  Some of the individual 2d spectra end up with all zeroes in the s2d output, while others show very blocky resampling with significant gaps.  The attached figure shows an example of both cases from the same exposure, which is a simulated MOS scene.  For the latter case, the problem might be fixed by using a finer grid of output points in the cross-dispersion direction (all spectra in this particular example only have 4 or 5 output pixels in that direction).  Is there a way to specify a different sampling than the default?

 

This exposure was processed using pipeline version 0.10.1a.dev79+gaf19b1ba.  The input file, calspec2 products, and MSA metafile can be found in /grp/jwst/wit4/nirspec/Muzerolle/sim_example/.

**Comment by stscijgbot** [comment]
Comment by James Davies: Thanks James.  Your pipeline version is from Aug 16.  On Aug 17, a big change to {{resample_spec}} was merged.  That hopefully fixed the issue of no data.

 

Which slit is blocky?  I notice that slit 51 has a very small dispersion size (8 pixels):

{{2018-09-12 15:49:10,387 - stpipe.Spec2Pipeline.extract_2d - INFO - Name of subarray extracted: 51}}
 {{2018-09-12 15:49:10,388 - stpipe.Spec2Pipeline.extract_2d - INFO - Subarray x-extents are: 2039 2047}}
 {{2018-09-12 15:49:10,388 - stpipe.Spec2Pipeline.extract_2d - INFO - Subarray y-extents are: 657 673}}

Would be nice to know which the problem slits were.

Thanks!

James

**Comment by stscijgbot** [comment]
Comment by James Muzerolle: Drat, I thought I was using the version that included those changes, I guess I missed it by one day!  I'll try updating my dev environment and reprocess.

 

The lower spectrum in the plot is for slit 3.  There are others exhibiting similar behavior, some with much bigger gaps in dispersion (e.g., slit 57).  Slit 51 is very short because most of the spectrum is projected onto the other detector; this is just the blue end that barely crosses onto NRS1.

**Comment by stscijgbot** [comment]
Comment by James Davies: Yes, I'm getting the same result.  Failure on slit 51.  If the last 4 pixels are reference pixels, should they be a part of the bounding box that defines where the data is?

Resample and drizzle should be able to handle NaNs, but this is probably an edge case in my own code.  I'll have a look.

One can remove slit 51 in the extract_2d file and run it through resample, and it seems to work fine.  No more slits with all zeros.  All the slits seem to be 4 or 5 pixels high when rectified.  Is this expected?

Slit 3 looks fine when I do a simple sum in the cross-dispersion direction and plot it up.  Blocky features seen in your plot are gone, and now it is smooth, though I will say it looks like the source is not centered in the extraction.  Hmm.

**Comment by stscijgbot** [comment]
Comment by James Davies: Fixed by PR #2644.

**Comment by stscijgbot** [comment]
Comment by James Muzerolle: Retested with version 0.13.0a.dev75+g326b6a2c.

The problem with the short slits crashing has been solved, and the slits that were originally coming out all zeroes are now producing something real.  However, the issue of the blocky resampling is still evident in all cases.  For example, slit 3 looks very similar to the previous result (the bottom panel of the original attached figure), just with more pixels in the cross-dispersion direction.  I'm not sure if this is a result of non-optimal sampling, or something with the algorithm itself.  How is the output pixel grid determined?  Is there a way to change it manually?

**Comment by stscijgbot** [comment]
Comment by James Muzerolle: Thanks for the info!  Your default wavelength sampling sounds reasonable.  I'm hoping to do some further studies once the initial implementation is done to try to optimize that; is there a way to change the sampling without having to modify the code, say by specifying it in a reference file?

In the meantime, you should be able to locate the exact slit position to which each pixel maps by looking at the slit_x and slit_y parameters.  These give the relative position in the slit in both dispersion and cross-dispersion directions, where [0,0] is the slit center and +/-0.5 are the edges, and can be calculated using the proper instrument model transforms.  I've glanced at your code, I can't tell if you have access to that information through the input data model?

**Comment by stscijgbot** [comment]
Comment by James Muzerolle: We're still testing this.  Based on our most recent runs, results have improved, but there are some issues that need more investigation.  This should be kept open for now.  I will try to post some results on a test ticket by tomorrow.

**Comment by stscijgbot** [comment]
Comment by Alicia Canipe: Thanks! Just checking older tickets.

**Comment by stscijgbot** [comment]
Comment by James Muzerolle: Reopened with new results from b7.4.

**Comment by stscijgbot** [comment]
Comment by Howard Bushouse: Lots of fixes included in [https://github.com/spacetelescope/jwst/pull/4766], which are tracked from JP-936. Closing this ticket. Any remaining issues can be added to JP-936 or a new ticket.
